woocommerce/woocommerce-ios
# [Orders with Coupons M4] Add discount to product business logic

Closes: #10202 


## Description

With this PR we implement the business logic to add a discount to a product. The EditableOrderViewModel creates a new product input with the discount and passes it to the order synchronizer, in a similar way we do for quantities. We therefore refactor the functions and classes involved to accomplish that.
We add a discount to a product by changing the item total (subtotal - discount). This is done in the `ProductInputTransformer`.
Apart from that, I didn't implement any mechanism to avoid higher discounts than the product price itself, as it doesn't make the order fail, and it's already allowed with (negative) fees. We can end up with a negative order total, but that's also allowed in WC Core.

## Testing instructions


1. Go to Orders.
2. Tap to Create a new order.
3. Add a product.
4. Tap on the product row.
5. Tap to add a discount. Tap on Done.
6. Go back to order details. See that the order total shows the discounted value. You can also check that once you create the order in wp-admin (See screenshots)

## Screenshots

<img src="https://github.com/woocommerce/woocommerce-ios/assets/1864060/29d29170-3180-4689-b5ff-90516d2bf6cd" width="375">

<img width="1376" alt="Screenshot 2023-07-11 at 14 23 37" src="https://github.com/woocommerce/woocommerce-ios/assets/1864060/edebfb20-8e28-40d5-aa13-d1b93828718d">


---
